common/cpp/src/google_smart_card_common/value_nacl_pp_var_conversion_unittest.cc

```diff
@@ -186,4 +188,269 @@ TEST(ValueNaclPpVarConversion, ArrayValue) {
   }
 }
 
+TEST(ValueNaclPpVarConversion, UndefinedPpVar) {
+  std::string error_message;
+  const optional<Value> value = ConvertPpVarToValue(pp::Var(), &error_message);
+  EXPECT_TRUE(error_message.empty());
+  ASSERT_TRUE(value);
+  EXPECT_TRUE(value->is_null());
+}
+
+TEST(ValueNaclPpVarConversion, NullPpVar) {
+  std::string error_message;
+  const optional<Value> value =
+      ConvertPpVarToValue(pp::Var(pp::Var::Null()), &error_message);
+  EXPECT_TRUE(error_message.empty());
+  ASSERT_TRUE(value);
+  EXPECT_TRUE(value->is_null());
+}
+
+TEST(ValueNaclPpVarConversion, BooleanPpVar) {
+  {
+    constexpr bool kBoolean = false;
+    std::string error_message;
+    const optional<Value> value =
+        ConvertPpVarToValue(pp::Var(kBoolean), &error_message);
+    EXPECT_TRUE(error_message.empty());
+    ASSERT_TRUE(value);
+    ASSERT_TRUE(value->is_boolean());
+    EXPECT_EQ(value->GetBoolean(), kBoolean);
+  }
+
+  {
+    constexpr bool kBoolean = true;
+    std::string error_message;
+    const optional<Value> value =
+        ConvertPpVarToValue(pp::Var(kBoolean), &error_message);
+    EXPECT_TRUE(error_message.empty());
+    ASSERT_TRUE(value);
+    ASSERT_TRUE(value->is_boolean());
+    EXPECT_EQ(value->GetBoolean(), kBoolean);
+  }
+}
+
+TEST(ValueNaclPpVarConversion, IntegerPpVar) {
+  constexpr int kInteger = 123;
+  std::string error_message;
+  const optional<Value> value =
+      ConvertPpVarToValue(pp::Var(kInteger), &error_message);
+  EXPECT_TRUE(error_message.empty());
+  ASSERT_TRUE(value);
+  ASSERT_TRUE(value->is_integer());
+  EXPECT_EQ(value->GetInteger(), kInteger);
+}
+
+TEST(ValueNaclPpVarConversion, FloatPpVar) {
+  constexpr double kFloat = 123.456;
+  std::string error_message;
+  const optional<Value> value =
+      ConvertPpVarToValue(pp::Var(kFloat), &error_message);
+  EXPECT_TRUE(error_message.empty());
+  ASSERT_TRUE(value);
+  ASSERT_TRUE(value->is_float());
+  EXPECT_DOUBLE_EQ(value->GetFloat(), kFloat);
+}
+
+TEST(ValueNaclPpVarConversion, StringPpVar) {
+  constexpr char kString[] = "foo";
+  std::string error_message;
+  const optional<Value> value =
+      ConvertPpVarToValue(pp::Var(pp::Var(kString)), &error_message);
+  EXPECT_TRUE(error_message.empty());
+  ASSERT_TRUE(value);
+  ASSERT_TRUE(value->is_string());
+  EXPECT_EQ(value->GetString(), kString);
+}
+
+TEST(ValueNaclPpVarConversion, ResourcePpVar) {
+  {
+    const optional<Value> value = ConvertPpVarToValue(pp::Var(pp::Resource()));
+    EXPECT_FALSE(value);
+  }
+
+  {
+    std::string error_message;
+    const optional<Value> value =
+        ConvertPpVarToValue(pp::Var(pp::Resource()), &error_message);
+    EXPECT_THAT(error_message, ::testing::HasSubstr("resource"));
```

Comment:
can we use `FormatPrintfTemplate(kUnsupportedPpVarTypeError, "resource");` here?

Reply:
Done. I originally wanted to avoid writing complex code for verifying those messages, but perhaps it made tests weirder indeed...